### ISSUE TITLE
request: logos for linux mint

### ISSUE THREAD

**Issue by brainchild0** [cockpit-project/cockpit#17878]
Would if be possible to enhance the Ubuntu package such that when used with Linux Mint, the corresponding logo is used, rather than the Ubuntu one? 

I am running Cockpit 276.1 under   Linux Mint 21.

**Comment by brainchild0** [comment]
The package used for Mint installations is taken from the main Ubuntu repository.

Linux Mint installations maintain the original Ubuntu logo image in  `/usr/share/pylmouth/ubuntu-logo.png`.

**Comment by brainchild0** [comment]
Perhaps the following changes might help lead to a clean resolution:
1. The file referenced as the logo image would have a vendor-neutral name (e.g. `logo.png`) at some location particular to the package.
2. A further package, given as dependency, would provide that referenced file, as symbolic link to the location currently currently given as the logo, `/usr/share/pylmouth/ubuntu-logo.png`.
3. *If desired, Mint may then provide its own version of the same package, in its repositories, which would be resolved by package managers on its systems, giving a different file, or link to a file elsewhere.*

**Comment by jelly** [comment]
I don't see why Cockpit has to change a working packaging approach which works for Arch, SUSE, Debian, Ubuntu, Fedora and RHEL. In the end your proposal still requires the Mint developers to act, which they could already have done by changing the provided package and rebuilding it to better integrate into Mint.

Note that, even if the logo was fixed in the Mint packaging, the CSS theming would still be Ubuntu specific.

**Comment by brainchild0** [comment]
@jelly, Mint uses Ubuntu packages for applications and much of the base, and adds its own repository for a few adjustments and the extra features. In order for broad compatibility and maintenance, it rarely supplies any package that would serve the same function as an existing Ubuntu one. As such, it is not exactly analogous in its composition to many other major distributions, such as the ones you have named.

Mint almost certainly lacks the resources to maintain a separate package. one that includes the full run time for Cockpit, simply due to differences in branding. However, by distributing the logo selection versus application functionality across two packages, Mint would be able to provide specific customization while still utilizing the application from the Ubuntu package base.

In this sense, the idea is less about Cockpit  "[having] to change a working packaging approach", and more about simply separating two different concerns, application versus logo, for better interoperability across various Ubuntu-based systems, including those not under the umbrella of Ubuntu branding.

Respecting style, yes, it seems that this problem is separate. Based on your comments, I understand that it may be solved simply by adding to the existing file hierarchy from the package, without introducing any conflicts. Would some further adjustment be required for selection of the different theme? 

Since the logo issue seemed the one that would require a more drastic solution, I thought it might be helpful to address in a separate proposal, as I have done.

**Comment by allisonkarlitskaya** [comment]
On fedora:

```
$ grep LOGO /usr/lib/os-release 
LOGO=fedora-logo-icon

$ find /usr/share/icons/ -name 'fedora-logo-icon.*'
/usr/share/icons/hicolor/16x16/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/22x22/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/24x24/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/256x256/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/32x32/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/36x36/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/48x48/apps/fedora-logo-icon.png
/usr/share/icons/hicolor/96x96/apps/fedora-logo-icon.png
```

**Comment by allisonkarlitskaya** [comment]
What I mean to say: this is probably a reasonable fallback for cases where we don't have built-in branding.  Then it would be on each distro to make sure they ship a reasonable icon and mention it in their `/usr/lib/os-release`.

**Comment by jelly** [comment]
@allisonkarlitskaya that is an excellent idea, but Debian, Ubuntu, Mint lack a `LOGO` and `COLORS`.

It turns out that we use branding based on `/etc/os-release` and as `Mint` has `ID_LIKE` ubuntu we grab the ubuntu theming on Mint. I feel that that approach is kinda broken and maybe we should stop looking at `ID_LIKE`.

**Comment by allisonkarlitskaya** [comment]
@jelly: the way to look up an icon name is defined by the [Icon Theme Specification](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-0.6.html#directory_layout).

I was proposing that we use `LOGO` if it's available, and have the existing set of upstream fallbacks for the case where it's not.  We could then make a policy of "we consider our upstream set of branding assets semi-deprecated and don't intend to expand them, use this other mechanism instead..."

**Comment by jelly** [comment]
> @jelly: the way to look up an icon name is defined by the [Icon Theme Specification](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-0.6.html#directory_layout).
> 
> I was proposing that we use `LOGO` if it's available, and have the existing set of upstream fallbacks for the case where it's not. We could then make a policy of "we consider our upstream set of branding assets semi-deprecated and don't intend to expand them, use this other mechanism instead..."

Right, we need to figure out if the LOGO is then in a discoverable place, so this requires some research.

**Comment by brainchild0** [comment]
It may be helpful to add a mechanism for discovering branding automatically based on conventions, rather than special cases of each distribution, but I would suggest an inversion of the current proposal. That is, I would suggest application logic that attempts automatic discovery after attempting to resolve resources from hard-coded locations, with the intention that package maintainers may populate those locations if the application logic is not compatible with the target system.

Ultimately, in my view, the objective should be to remove completely from the core application any direct awareness of any distributions, and to leave any required adjustments as features of a package.

It is not clear whether such an approach completely would remove the need for allowing LM to override certain resources in a separate package, but the idea of separating the packaging of core application from site branding  is in general a  good design, allowing more flexibility and robustness for the future.

The following general outline of a discovery procedure may suggest some thoughts for a new approach:
1. Using conventions for *Icon Theme Specification*, attempt to resolve image resource with name `cockpit-logo`.
2. If unsuccessful, similarly attempt to resolve resource `cockpit-${DIST}-logo`, where "`${DIST}`" shall be interpolated to the value of the field `ID` from  `/etc/os-release`.
3. If unsuccessfully, attempt the same, where "`${DIST}`" shall  be interpolated as each word, in iteration until successful,   in the field `ID_LIKE` from  `/etc/os-release`.
6. If unsuccessful, similarly attempt to resolve resource with the name from the value of the field `LOGO` in `/etc/os-release`.


In some cases, the last rule would be sufficient, but the others give package maintainers latitude to address differences across distributions.

**Comment by brainchild0** [comment]
If Cockpit is to use the  Icon Theme Specification, then it may help to invoke the functionality provided by [*PyGObject*](https://www.gtk.org/docs/language-bindings/python/), in order to avoid a need for directly implementing the logic. Even though Cockpit is not a GTK application, the library, including the Python bindings, are widely distributed, and the functionality may be accessed without reliance on any display resources.

The [`IonTheme`](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/IconInfo.html) class has class method [`choose_icon`](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/IconTheme.html#Gtk.IconTheme.choose_icon). It returns an instance of [`IconInfo`](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/IconInfo.html),   which offers method [`get_filename`](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/IconInfo.html#Gtk.IconInfo.get_filename). 

With these calls, it would be possible to resolve the file location corresponding to an icon resource by name.

**Comment by martinpitt** [comment]
Stalled, and our team doesn't have the resources to look into providing a Mint logo. PRs appreciated of course!
